### ISSUE TITLE
Meson coverage fails while using clang compiler with gcov

### ISSUE THREAD

**Issue by abhay2703** [mesonbuild/meson#6691]
**Describe the bug**
Meson coverage fails while using clang compiler with gcov
Error
<filename>cpp.gcno:version '402*', prefer '408R'

gcov: out of memory allocating 16158259920 bytes after a total of 135168 bytes
geninfo: ERROR: GCOV failed for <filename>.cpp.gcda!
Traceback (most recent call last):
  File "/usr/local/bin/meson", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 226, in main
    return run(sys.argv[1:], launcher)
  File "/usr/local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 215, in run
    return run_script_command(args[1], args[2:])
  File "/usr/local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 163, in run_script_command
    return module.run(script_args)
  File "/usr/local/lib/python3.6/site-packages/mesonbuild/scripts/coverage.py", line 147, in run
    options.log_dir)
  File "/usr/local/lib/python3.6/site-packages/mesonbuild/scripts/coverage.py", line 72, in coverage
    '--rc', 'lcov_branch_coverage=1',
  File "/usr/lib64/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['lcov', '--directory', '<buildDIR>', '--capture', '--output-file', '<buildDIR>/meson-logs/coverage.info.run', '--no-checksum', '--rc', 'lcov_branch_coverage=1']' returned non-zero exit status 29.
ninja: build stopped: subcommand failed.

**Expected behavior**
Coverage should work independent of compiler (gcc or clang)

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? - native build
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) - CentOS Linux release 7.7.1908
* what Python version are you using e.g. 3.8.0 - Python 3.6.8
* what `meson --version`  - 0.53.1
* what `ninja --version` if it's a Ninja build - 1.7.2

**Comment by nirbheek** [comment]
The most significant part of the log is: 

```
Error
cpp.gcno:version '402*', prefer '408R'

gcov: out of memory allocating 16158259920 bytes after a total of 135168 bytes
```

Looks like a bug in gcov?

**Comment by abhay2703** [comment]
Hi Nirbheek,

Yes this is correct.
Since clang is not compatible with gcov, this can be resolved using
llvm-cov with clang.

On my local copy I have found a workaround by setting gcov-tool as llvm-cov
in coverage.py file.
 So can this be resolved from your end?

Regards,
Abhay Jain

**Comment by abhay2703** [comment]
In coverage.py, I made the following changes:
 67             subprocess.check_call([lcov_exe,
 68                                    '--directory', build_root,
 69                                    '--base-directory', build_root,
 70                                    **'--gcov-tool', '../llvm.sh',**
 71                                    '--capture',
 72                                    '-o', run_tracefile,
 73                                    ])

And the llvm.sh content is as follows:
 1 #!/bin/bash
 2 exec llvm-cov gcov "$@"

**Comment by abhay2703** [comment]
> The most significant part of the log is:
> 
> ```
> Error
> cpp.gcno:version '402*', prefer '408R'
> 
> gcov: out of memory allocating 16158259920 bytes after a total of 135168 bytes
> ```
> 
> Looks like a bug in gcov?

Since gcov/lcov and clang are not compatible, shouldn't this be the responsibility of meson to search for some compatible tool with clang for coverage?
Or meson should give option to user to select coverage tool using some argument!?

**Comment by eli-schwartz** [comment]
This was fixed by #7196
